index.js

```diff
@@ -1,43 +1,74 @@
 'use strict';
-const mem = require('mem');
 const mimicFn = require('mimic-fn');
+const mapAgeCleaner = require('map-age-cleaner');
 
-const memoizedFunctions = new WeakMap();
+const cacheStore = new WeakMap();
 
 const pMemoize = (fn, {cachePromiseRejection = false, ...options} = {}) => {
+	const {maxAge, cacheKey} = options;
 	const cache = options.cache || new Map();
-	const cacheKey = options.cacheKey || (([firstArgument]) => firstArgument);
 
-	const memoized = mem(fn, {
-		...options,
-		cache,
-		cacheKey
-	});
+	if (!isNaN(maxAge)) {
+		mapAgeCleaner(cache);
+	}
+
+	const memoized = function (...arguments_) {
+		const key = cacheKey ? cacheKey(arguments_) : arguments_[0];
 
-	const memoizedAdapter = function (...arguments_) {
-		const cacheItem = memoized.apply(this, arguments_);
+		const cacheItem = cache.get(key);
+		if (cacheItem) {
+			return cacheItem.data;
+		}
 
-		if (!cachePromiseRejection && cacheItem && cacheItem.catch) {
-			cacheItem.catch(() => {
-				cache.delete(cacheKey(arguments_));
+		const result = fn.apply(this, arguments_);
+
+		let resultError = null;
+		const resultFinally = () => {
+			cache.set(key, {
+				data: result,
+				maxAge: maxAge ? Date.now() + maxAge : Number.POSITIVE_INFINITY
 			});
+
+			if (!cachePromiseRejection && resultError) {
+				cache.delete(key);
+			}
+		};
+
+		if (result && result instanceof Promise) {
```

Comment:
`instanceof` is not a good way to check for promises.

```suggestion
		if (result && typeof result.then === 'function') {
```

Reply:
Why are you checking for a promise at all? You can just use `Promise.resolve()`.

Reply:
Because `Promise.resolve()` returns a promise and I think it is better to keep the function working as it is now. That means when it gets a non-promise returning function it won't change it to a promise returrning function.

Reply:
The docs are pretty clear about it only accepting promise-returning functions though, even though we didn't actually validate that:

> Promise-returning or async function to be memoized.

Reply:
Yes, you're right. I guess it's a matter of taste, I didn't want to change the behaviour if not necessary.

But if you look at it as a "bug", then it would make sense to change now.

Reply:
Yeah, I see it as a bug.

Reply:
I've used the  `Promise.resolve()` now.